### PR TITLE
do both adler32 and crc checksum calculation in one file read

### PR DESCRIPTION
**Pull request by hufnagel**
To improve performance and reduce disk IO on the worker node calculate the adler32 and crc checksums in one file read.

### CODE REVIEW

**Comment by hufnagel** [comment]
New pull request as it completely changes the goal of the changes. @ticoann , I still removed the crc checksum from wmstats, as IMO we should only deal with adler32 everywhere as we can (crc checksum will just stay around for legacy reasons.

But I can remove the wmstats changes if that makes validating and eventually merging this easier...

**Comment by hufnagel** [comment]
@PerilousApricot , also would appreciate a look at my checksum calculation code. Its a (much) simpler version than you presented in the other issue.

Also, please note that cksum.stdin.write is highly discouraged by the python documentation, but is the only way this works. The suggested cksum.communicate() cannot deal with passing input data in chunks at a time.

**Comment by hufnagel** [comment]
Also, do not merge yet, this is untested !

**Comment by hufnagel** [comment]
Also included passing NOTSET as adler32 or md5 checksum to DBS if they are not defined. Instead don't pass anything to DBS.

**Comment by ticoann** [comment]
For WMStats, it doesn't really matter. It is just a display purpose. If information is not needed we don't have there. Validating and merging is the same unless this need to be merged before May production deployment. But if this has to be merged before that we can do that as well.

**Comment by hufnagel** [comment]
Patch has been tested in a Tier0 replay. I looked at the DBS file inject records, it only passed the adler32 and crc checksums now, not md5 anymore (and the injection still works). I also took the checksums passed to DBS and compared them to interactively calculated checksums from the same file copied from EOS to local disk.

**Comment by hufnagel** [comment]
Lets leave the wmstats in place then. We keep crc checksums, because a) we save ourselves some headache with DBS and b) we get them basically for free now. They are still there, but we don't use them prominently anymore. Who knows, maybe they'll go away in DBS4 :-).

As for wmstats, as the checksum is still there, the change can be merged whenever. Its not critical that the wmstats change is deployed at the same time as the other changes, this can happen in any order.

**Comment by hufnagel** [comment]
@ericvaandering @ticoann @PerilousApricot , please review/comment.

**Comment by PerilousApricot** [comment]
:+1:  - Sucks that we have to spawn off a process for checksums, but like you noticed, there's something different between the command line and what zlib does :/. Rest of it looks good to me, esp. if you verified that it all looks the same interactively

**Comment by hufnagel** [comment]
Btw, my interactive tests were mostly for the crc checksum calculated by cksum. If we do get adler32 wrong at the end of the job stageout will fail immediately because stageout actually checks that file written to EOS has the samer adler32 as what we pass with the file back to the agent.

**Comment by ericvaandering** [comment]
So is this the proposed end state for this issue or do we move forward with the DBS issue and simplify this code in June/July?

File: src/python/WMCore/Algorithms/BasicAlgos.py [hunk 1]
```diff
@@ -10,7 +10,48 @@
...
+    The cksum UNIX command line tool implements a CRC32 checksum that is
+    different than any of the python algorithms, therefore open cksum
+    in a subprocess and feed it the same chunks of data that are used
+    to calculate the adler32 checksum.
+
+    """
+    adler32Checksum = 1 # adler32 of an empty string
+    cksumProcess = subprocess.Popen("cksum", stdin = subprocess.PIPE, stdout = subprocess.PIPE)
+
+    f = open(filename, 'rb')
```
Comment (ericvaandering) [inline src/python/WMCore/Algorithms/BasicAlgos.py]: You can use a context manager here (with open as f)

**Comment by hufnagel** [comment]
Well, yes, removing the crc checksum from the WMCore core will simplify the code somewhat (removing like 10 lines of code in one file and a line of code each from two other files). But there really isn't much of performance benefit anymore (a few seconds for the crc algorithm to calculate from data already in memory) and with all these files in DBS already having the CRC checksum, there is an argument to be made to keep supplying it for consistency.

I think it depends on what happens on the DBS side, which is largely driven by what the policy is. Mandatory crc checksum and other checksums optional doesn't make much sense as a policy, but it is what is implemented. In effect we always pass both from WMAgent (not sure what CRAB3 does, they should probably also get these code updates...).

More a question for you though, do we want to spend time on what has stopped being a performance issue and has moved to be a policy clarification and implementation cleanup to be more consistent with a possible new policy ?

Even if we decide we still want to move forward on the DBS side, I would suggest to validate/test/merge this code update as-is and then make the other changes later once we know what we want to do on the DBS side.

File: src/python/WMCore/Algorithms/BasicAlgos.py (hunk 1 above)
Comment (hufnagel) [inline src/python/WMCore/Algorithms/BasicAlgos.py]: Just read up on it. No functional difference and the old style syntax isn't deprecated or anything like that either. There are no benefits for a simple example like this here, although I can see it being useful if what you ar doing in the file processing section is more complex.

**Comment by hufnagel** [comment]
@ericvaandering , update the file open and loop construct. Works interactively, although I am also running another Tier0 replay with it.

**Comment by hufnagel** [comment]
Replay worked with the latest code as well, spot check on a produced file and the cksum passed to DBS shows it to be identical to the same cksum calculated from another file copy from EOS.

**Comment by ticoann** [comment]
@hufnagel, Dirk, this is ready to be merged, right?

**Comment by hufnagel** [comment]
From my side definitely. I am using it in all Tier0 replays I am doing.

**Comment by ericvaandering** [comment]
Merge it. Ignore my Jenkins prototype which is this "Waiting to hear" thing

**Pull request merged**